### ISSUE TITLE
akka-http client PoolGateway stuck in spinloop, eventually using 100% CPU

### ISSUE THREAD

**Issue by nkvoll** [akka/akka#20071]
Unfortunately, I don't currently know *why* or *how* it happens, but it is possible for the `state` in a `PoolGateway` to be set to a `NewIncarnation(newGatewayFuture)`, where `newGatewayFuture` is a `FastFuture.successful(..)` of it self.

In other words, at https://github.com/akka/akka/blob/master/akka-http-core/src/main/scala/akka/http/impl/engine/client/PoolGateway.scala#L57, `this == previousIncarnation`. By attaching a debugger to the process I've added a breakpoint that doesn't suspend, but only logs `this` and `previousIncarnation`, and it keeps logging the following:

    this: akka.http.impl.engine.client.PoolGateway@79fcbcef - previous: akka.http.impl.engine.client.PoolGateway@79fcbcef
    this: akka.http.impl.engine.client.PoolGateway@1df76436 - previous: akka.http.impl.engine.client.PoolGateway@1df76436
    this: akka.http.impl.engine.client.PoolGateway@516279e7 - previous: akka.http.impl.engine.client.PoolGateway@516279e7
    this: akka.http.impl.engine.client.PoolGateway@79fcbcef - previous: akka.http.impl.engine.client.PoolGateway@79fcbcef
    this: akka.http.impl.engine.client.PoolGateway@1df76436 - previous: akka.http.impl.engine.client.PoolGateway@1df76436
    this: akka.http.impl.engine.client.PoolGateway@516279e7 - previous: akka.http.impl.engine.client.PoolGateway@516279e7

I'm guessing it's a race condition somewhere as it does not happen often, but when it does, the thread that's running calling the method will spin.

Observations from canary-ing it in production is that it can take a long time before it occurs (last time for me it took 90 hours under moderate workload for it to happen to the first thread), but just a few hours until there were four requests spinning (on a server with four cpu cores.). Hopefully that observation may help in finding the cause.

I'm seeing this using the following version(s):

    "com.typesafe.akka:akka-actor_2.11:2.4.2",
    "com.typesafe.akka:akka-http-core_2.11:2.4.2",
    "com.typesafe.akka:akka-http-experimental_2.11:2.4.2",
    "com.typesafe.akka:akka-parsing_2.11:2.4.2",
    "com.typesafe.akka:akka-stream_2.11:2.4.2",

**Comment by nkvoll** [comment]
Is there anything else I can / should try to provide to assist in finding a cause / debugging further on the canary server in question?

**Comment by nkvoll** [comment]
This seems to be a race condition that can happen to any request using the cached connection pools just as the current incarnation of the pool is being shutdown.

Looking at the code from https://github.com/akka/akka/blob/master/akka-http-core/src/main/scala/akka/http/scaladsl/Http.scala#L570-L582:

```
        val whenShuttingDown = Promise[Done]()
        val gateway =
          try new PoolGateway(setup, whenShuttingDown)
          catch {
            case NonFatal(e) ⇒
              hostPoolCache.remove(setup)
              gatewayPromise.failure(e)
              throw e
          }
        val fastFuture = FastFuture.successful(gateway)
        hostPoolCache.put(setup, fastFuture) // optimize subsequent gateway accesses
        gatewayPromise.success(gateway) // satisfy everyone who got a hold of our promise while we were starting up
        whenShuttingDown.future.onComplete(_ ⇒ hostPoolCache.remove(setup, fastFuture))(fm.executionContext)

```

Isn't there a simple race condition here? The `hostPoolCache` is only updated some time after `whenShuttingDown`  is completed, while the code in https://github.com/akka/akka/blob/master/akka-http-core/src/main/scala/akka/http/impl/engine/client/PoolGateway.scala#L82:

```
        if (state.compareAndSet(x, IsShutdown(shutdownCompletedPromise.future))) {
          shutdownStartedPromise.success(Done) // trigger cache removal
          ref ! PoolInterfaceActor.Shutdown
          shutdownCompletedPromise.future
        } else shutdown() // CAS loop (not a spinlock)
````

Sets the `IsShutdown` state to use the same future, which is matched at https://github.com/akka/akka/blob/master/akka-http-core/src/main/scala/akka/http/impl/engine/client/PoolGateway.scala#L63. That code assumes that it can call `Http().cachedGateway` to get the next `PoolGateway`.. but that isn't the case:

This `shutdownCompletedPromise.future` has two actions (one added by `onComplete`, the other in the `IsShutdown` state match as the flatMap), using different threads with no guarantee of their ordering.

The former code should probably be updated in one way or another to give the `PoolGateway` a reference to a future that already has cleanup performed, so the ordering is guaranteed.

**Comment by nkvoll** [comment]
I've attempted to provide a fix for the issue in https://github.com/akka/akka/pull/20190. Could anyone take a look?

**Comment by ktoso** [comment]
This is now fixed via #20082 and will be part of Akka 2.4.4
